pep-0695.rst

```diff
@@ -0,0 +1,1103 @@
+
+
+Compiler Changes
+----------------
+
+The compiler maintains a list of "active type parameters" as it recursively
+generates byte codes for the program. Consider the following example.
+
+::
+
+    class Outer[K, V]:
+        # Active type parameters are K and V
+
+        class Inner[T]:
+            # Active type parameters are K, V, and T
+
+            def method[M](self, a: M) -> M:
+                # Active type parameters are K, V, T, and M
+                ...
+
+An active type parameter symbol cannot be used for other purposes within
+these scopes. This includes local parameters, local variables, variables
+bound from other scopes (nonlocal or global), or other type parameters. An
+attempt to reuse a type parameter name in one of these manners results in
+a syntax error.
+
+::
+
+    class ClassA[K, V]:
+        class Inner[K]: # Syntax error: K already in use as type parameter
+            ...
+
+    class ClassB[K, V]:
+        def method(self, K): # Syntax error: K already in use as type parameter
+            ...
+
+    class ClassC[T, T]: # Syntax error: T already in use as type parameter
+        ...
+
+    def func1[T]():
+        ...
+
+
+A type parameter is considered "active" when compiling the arguments for
+a class declaration, the type annotations for a function declaration, and
+the right-hand expression in a type alias declaration. Type parameters are
+not considered "active" when compiling the default argument expressions for
+a function declaration or decorator expressions for classes or functions.
+
+::
+
+    T = list
+
+    @decorator(T) # T in decorator refers to outer variable
+    class ClassA[T](Base[T], metaclass=Meta[T]) # T refers to type parameter
+        ...
+
+    @decorator(T) # T in decorator refers to outer variable
+    def func1[T](a: list[T]) -> T: # T refers to type parameter
+        ...
+
+    def func2[T](a = T): # T in default refers to outer variable
+        ...
+
+
+When a type parameter is referenced, the compiler generates the byte codes
+to construct a ``typing.TypeParameter`` object. This new class can be thought
+of as a proxy for ``TypeVar``, ``TypeVarTuple`` or ``ParamSpec``. It has a
+name and ``args`` and ``kwargs`` properties (required for ``ParamSpec``) so
+it can be used within a type expression at runtime. It also implements a
+``__ror__`` method so it can be used in unions, an ``__iter__`` method so
+it can be unpacked (needed for ``TypeVarTuple``) and a ``__repr__`` method
+so its name can be printed. It does not allow for runtime differentiation
+between a ``TypeVar``, ``TypeVarTuple`` or ``ParamSpec``, nor does it
+allow for runtime introspection of the upper bound, constraints, or
+variance.
+
+If a bound expression is provided for a type parameter, it is validated for
+syntax but is not evaluated at runtime. That is, no byte codes are generated
+for the bound expression. Type checkers should evaluate its type statically.
```

Comment:
That's a problem for runtime type checkers, because they'd have no way of getting the bound.

Ideally I'd want the bound to be accessible as a code object that can be evaluated on demand, like PEP 649's design for `__annotations__`.

Reply:
Let's discuss this in the typing-sig. Evaluating the bound here won't be possible with my proposed implementation. Implementations that would enable this are extremely complex and/or impose a high runtime cost. We should discuss how important it is to runtime type checkers that bounds and constraints be checked. We should also explore whether there are other ways for runtime type checkers to get access to the type parameters and the detailed information within them.